security/elastic.py

```diff
@@ -0,0 +1,106 @@
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import json
+import logging
+
+import elasticsearch
+from elasticsearch import helpers
+
+from security import base
+
+INDEX = "ms_security_"
+
+
+class StorageException(Exception):
+    pass
+
+
+class Backend(object):
+    """Represent elasticsearch backend.
+
+    usage:
+        b = elastic.Backend(urls=["url1", "url2"])
+        issues = b.get_issues("Region1", "IssueType1,IssueType2")
+        b.begin()
+        b.store_issue(Issue(...))
+        b.commit("region1")
+    """
+
+    def __init__(self, urls, **kwargs):
+        self.index = INDEX
+        self.es = elasticsearch.Elasticsearch(urls)
+        self.body = ""
+
+    def get_issues(self, region, types=None, discovered_after=None):
+        query = {
+            "query": {
+                "bool": {
+                    "must_not": {
+                        "exists": {
+                            "field": "resolvedAt",
+                        },
+                    },
+                },
+            }
+        }
+        if discovered_after:
+            query["query"]["bool"]["must"] = {
+                "range": {
+                    "discoveredAt": {"gt": discovered_after.isoformat("T")}
+                }
+            }
+        issues = []
+        try:
+            for i in helpers.scan(self.es, index=self.index + region,
+                                  doc_type=types, query=query, scroll="1m"):
+                kwargs = i["_source"]
+                kwargs["discovered_at"] = kwargs["discoveredAt"]
+                kwargs["confirmed_at"] = kwargs["confirmedAt"]
+                kwargs["resolved_at"] = kwargs.get("resolvedAt")
+                issues.append(base.Issue(i["_type"], **kwargs))
+            return issues
+        except elasticsearch.NotFoundError:
+            return []
+
+    def begin(self):
+        if self.body:
+            logging.warning("Unsaved data %s", self.body)
+        self.body = ""
+
+    def commit(self, region):
+        try:
+            self.es.indices.create(index=self.index + region)
+        except elasticsearch.RequestError as ex:
+            if ex.error != "index_already_exists_exception":
+                raise
+        retval = self.es.bulk(self.body, self.index + region)
+        self.body = ""
+        try:
+            for item in retval["items"]:
+                if 200 > item["index"]["status"] > 201:
```

Comment:
Seems like this can't be true maybe, first of all there is no number between 200 and 201, and 200 is always < 201 which means this is dead code.